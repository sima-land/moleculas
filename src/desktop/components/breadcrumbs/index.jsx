import React, { useState, useRef, useEffect } from 'react';
import CSSTransition from 'react-transition-group/CSSTransition';
import Popup from '@dev-dep/ui-nucleons/popups/popup';
import on from '@dev-dep/ui-nucleons/helpers/on';
import classnames from 'classnames/bind';
import prop from 'lodash/fp/prop';
import DownSVG from './square-arrow-down.svg';
import styles from './breadcrumbs.scss';

const cx = classnames.bind(styles);

const DEFAULTS = {
  getName: prop('name'),
  getItemURL: prop('url'),
  getItemSiblings: prop('items'),
  isActiveItem: prop('isActive'),
  hasSiblings: breadcrumb => Array.isArray(breadcrumb?.items) && breadcrumb.items.length > 0,
};

/**
 * Список "хлебных крошек".
 * @param {Object} props Свойства компонента.
 * @param {Array} props.items Список "Хлебных крошек".
 * @param {Function} [props.getItemName] Вернёт имя элемента списка (по умолчанию свойство "name").
 * @param {Function} [props.getItemURL] Вернёт URL элемента списка (по умолчанию свойство "url").
 * @param {Function} [props.getItemSiblings] Вернёт список дочерних ссылок (по умолчанию свойство "items").
 * @param {Function} [props.isItemActive] Вернёт true если элемент представляет активный раздел.
 * @param {Function} [props.hasItemSiblings] Вернёт true если имеются соседние категории.
 * @param {Function} [props.onItemSiblingsShow] Будет вызвана при открытии списка соседних элементов.
 * @param {string} [props.'data-testid'] Идентификатор для систем автоматизированного тестирования.
 * @return {ReactElement} Элемент.
 */
export const Breadcrumbs = ({
  items,
  getItemName = DEFAULTS.getName,
  getItemURL = DEFAULTS.getItemURL,
  getItemSiblings = DEFAULTS.getItemSiblings,
  isItemActive = DEFAULTS.isActiveItem,
  hasItemSiblings = DEFAULTS.hasSiblings,
  onItemSiblingsShow,
  'data-testid': dataTestId,
}) => (
  <ul className={cx('breadcrumbs')} data-testid={dataTestId}>
    {Array.isArray(items) && items.map((breadcrumb, index) => (
      <Breadcrumb
        key={index}
        title={getItemName(breadcrumb)}
        href={getItemURL(breadcrumb)}
        canOpen={hasItemSiblings(breadcrumb)}
        siblings={getItemSiblings(breadcrumb)}
        active={isItemActive(breadcrumb)}
        getSiblingName={getItemName}
        isActiveSibling={isItemActive}
        getSiblingHref={getItemURL}
        onSiblingsShow={() => onItemSiblingsShow?.(breadcrumb)}
        data-testid={`breadcrumb:${index}`}
      />
    ))}
  </ul>
);

/**
 * Элемент списка "хлебных крошек".
 * @param {Object} props Свойства.
 * @return {ReactElement} Элемент.
 */
const Breadcrumb = ({
  title,
  href,
  siblings,
  active,
  getSiblingName,
  isActiveSibling,
  getSiblingHref,
  onSiblingsShow,
  canOpen,
  'data-testid': dataTestId,
}) => {
  const popupRef = useRef();
  const openerRef = useRef();
  const hasSiblings = siblings?.length > 0;
  const breadcrumbName = active
    ? (
      <span className={cx('active')}>{title}</span>
    ) : (
      <a
        className={cx('link')}
        href={href}
        children={title}
      />
    );

  const [withPopup, togglePopup] = useSiblingsPopup(popupRef, openerRef);

  return (
    <li data-testid={dataTestId}>
      {breadcrumbName}
      {canOpen && (
        <span ref={openerRef}>
          <DownSVG
            data-testid='breadcrumb:siblings-opener'
            className={cx('toggle-icon')}
            onClick={() => {
              togglePopup(true);
              onSiblingsShow?.();
            }}
          />
        </span>
      )}
      {hasSiblings && (
        <CSSTransition
          in={withPopup}
          timeout={200}
          unmountOnExit
          classNames={{
            enter: cx('fade-enter'),
            enterActive: cx('fade-enter-active'),
            exit: cx('fade-exit'),
            exitActive: cx('fade-exit-active'),
          }}
        >
          <Popup
            additionalClass={cx('popup')}
            ref={popupRef}
          >
            {breadcrumbName}
            <DownSVG
              data-testid='breadcrumb:siblings-closer'
              className={cx('toggle-icon', 'close')}
              onClick={() => togglePopup(false)}
            />
            <ul className={cx('siblings-list')}>
              {siblings.map((siblingItem, siblingIndex) => {
                const isActive = isActiveSibling(siblingItem);
                const siblingName = getSiblingName(siblingItem);

                return (
                  <li
                    key={siblingIndex}
                    className={cx('sibling-item')}
                    data-testid={`breadcrumbs-item-sibling:${siblingIndex}`}
                  >
                    {isActive ? (
                      <span className={cx('disabled')}>{siblingName}</span>
                    ) : (
                      <a
                        className={cx('link')}
                        color='dark-gray'
                        href={getSiblingHref(siblingItem)}
                        children={siblingName}
                      />
                    )}
                  </li>
                );
              })}
            </ul>
          </Popup>
        </CSSTransition>
      )}
    </li>
  );
};

/**
 * Хук работы с попапом соседних элементов "крошки".
 * @param {Object} popupRef Ref попапа.
 * @param {Object} openerRef Ref открывающего элемента.
 * @return {Function} Функция-обработчик монтирования компонента.
 */
const useSiblingsPopup = (popupRef, openerRef) => {
  const [withPopup, togglePopup] = useState(false);

  useEffect(() => {
    const offClick = on(window, 'click', ({ target }) => {
      target

        // если клик произошел не по узлу иконки открытия и не внутри popup'а
        && [popupRef.current, openerRef.current].every(node => node && node !== target && !node.contains(target))
        && togglePopup(false);
    });

    const offKeyDown = on(window, 'keydown', ({ code, keyCode }) => {
      (code === 'Escape' || keyCode === 27) && togglePopup(false);
    });

    return () => {
      offClick();
      offKeyDown();
    };
  }, []);

  return [withPopup, togglePopup];
};
