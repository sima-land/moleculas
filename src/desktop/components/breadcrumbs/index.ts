export { type Breadcrumb, type BreadcrumbsProps, Breadcrumbs } from './breadcrumbs';
