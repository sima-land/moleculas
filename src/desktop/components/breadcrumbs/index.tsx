import React, { useEffect, useState, useRef } from 'react';
import CSSTransition from 'react-transition-group/CSSTransition';
import { Plate } from '@sima-land/ui-nucleons/plate';
import { Link } from '@sima-land/ui-nucleons/link';
import { times } from 'lodash';
import classnames from 'classnames/bind';
import styles from './breadcrumbs.module.scss';
import DownSVG from './square-arrow-down.svg';
import on from '@sima-land/ui-nucleons/helpers/on';

export interface Sibling {
  name: string;
  url?: string;
  [key: string]: any;
}

export interface Breadcrumb extends Sibling {
  siblings?: {
    state: 'initial' | 'fetching' | 'ready';
    data?: Sibling[];
  };
}

export interface BreadcrumbsProps {
  /** Список хлебных крошек. */
  items: Breadcrumb[];

  /** Получив хлебную крошку должна вернуть true если крошка выбрана. */
  isSiblingActive?: (sibling: Sibling) => boolean;

  /** Сработает при открытии попапа со смежными ссылками. */
  onSiblingsPopupOpen?: (item: Breadcrumb) => void;

  /** Идентификатор для систем автоматизированного тестирования. */
  'data-testid'?: string;
}

const cx = classnames.bind(styles);

/**
 * Список "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
export const Breadcrumbs = ({
  items,
  onSiblingsPopupOpen,
  isSiblingActive,
  'data-testid': testId = 'breadcrumbs',
}: BreadcrumbsProps) => (
  <nav className={cx('root')} data-testid={testId}>
    {items.map((breadcrumb, index) => (
      <Breadcrumb
        key={index}
        data={breadcrumb}
        onSiblingsPopupOpen={() => onSiblingsPopupOpen?.(breadcrumb)}
        isSiblingActive={sibling => Boolean(isSiblingActive?.(sibling))}
      />
    ))}
  </nav>
);

/**
 * Элемент списка "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
const Breadcrumb = ({
  data,
  onSiblingsPopupOpen,
  isSiblingActive,
}: {
  data: Breadcrumb;
  isSiblingActive: (sibling: Breadcrumb) => boolean;
  onSiblingsPopupOpen: () => void;
}) => {
  const popupRef = useRef<HTMLDivElement>(null);

  const [withPopup, togglePopup] = useState<boolean>(false);

  // с useOutsideClick не работает после обновления React до 17 версии, пока так
  useEffect(() => {
    if (withPopup) {
      return on(window, 'mousedown', togglePopup.bind(null, false));
    }
  }, [withPopup]);

  const selfName = (
    <Link color='gray87' href={data.url}>
      {data.name}
    </Link>
  );

  return (
    <div data-testid='breadcrumb' className={cx('breadcrumb')}>
      {selfName}

      {data.siblings && (
        <>
          <DownSVG
            data-testid='breadcrumb:siblings-opener'
            className={cx('toggle-icon')}
            onClick={() => {
              togglePopup(true);
              onSiblingsPopupOpen();
            }}
          />

          <CSSTransition
            in={withPopup}
            timeout={200}
            unmountOnExit
            classNames={{
              enter: cx('fade-enter'),
              enterActive: cx('fade-enter-active'),
              exit: cx('fade-exit'),
              exitActive: cx('fade-exit-active'),
            }}
          >
            <Plate ref={popupRef} shadow='z3' rounds='m' className={cx('popup')}>
              <div className={cx('popup-header')}>
                {selfName}
                <DownSVG
                  data-testid='breadcrumb:siblings-closer'
                  className={cx('toggle-icon', 'close')}
                  onClick={() => togglePopup(false)}
                />
              </div>

              <ul className={cx('siblings')}>
                {data.siblings.state === 'fetching' && (
                  <>
                    {times(8).map(index => (
                      <div key={index} className={cx('sibling-placeholder')} />
                    ))}
                  </>
                )}

                {data.siblings.state === 'ready' && (
                  <>
                    {data.siblings.data?.map((sibling, index) => (
                      <li key={index} className={cx('sibling')} data-testid='breadcrumb-sibling'>
                        {isSiblingActive(sibling) ? (
                          <span className={cx('active')}>{sibling.name}</span>
                        ) : (
                          <Link
                            color='gray87'
                            className={cx('link')}
                            href={sibling.url}
                            children={sibling.name}
                          />
                        )}
                      </li>
                    ))}
                  </>
                )}
              </ul>
            </Plate>
          </CSSTransition>
        </>
      )}
    </div>
  );
};
