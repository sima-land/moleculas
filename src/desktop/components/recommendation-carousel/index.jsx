import React, { useRef, useEffect } from 'react';
import Carousel from '@dev-dep/ui-nucleons/carousel';
import RecommendedItem from './recommended-item';
import CarouselButton from './recommendation-carousel-button';
import withInViewportObserver from '@dev-dep/ui-nucleons/with-in-viewport-observer';
import isFunction from 'lodash/isFunction';
import { isNonEmptyArray } from '../../../common/helpers/is-non-empty';
import classes from './recommendation-carousel.scss';
import classnames from 'classnames/bind';
import Types from 'prop-types';

const cx = classnames.bind(classes);

/**
 * Компонент карусели рекомендованных товаров.
 * @param {Object} props Свойства компонента.
 * @param {string} props.title Заголовок карусели.
 * @param {Array} props.items Массив с рекомендованными товарами.
 * @param {Function} props.onItemClick Обработчик клика по товару.
 * @param {Function} props.onInViewport Функция, вызываемая при попадании компонента в поле видимости.
 * @param {Function} props.onRequest Функция, инициализирующая загрузку рекомендаций.
 * @param {Function} props.addObserve Функция подписки на Intersection Observer.
 * @param {Function} props.addObserveWithMargin
 * Функция подписки на Intersection Observer с увеличенным пределом попадания в область видимости.
 * @return {ReactElement} Компонент карусели рекомендаций.
 */
export const RecommendationCarousel = ({
  title,
  items,
  onItemClick,
  onInViewport,
  onRequest,
  addObserve,
  addObserveWithMargin,
}) => {
  const sectionRef = useRef();
  const emptyRef = useRef();

  // Отправляем статистку в ОКО, когда компонент попадает в зону видимости
  useEffect(() => {
    isFunction(addObserve)
    && isFunction(onInViewport)
    && sectionRef.current
    && addObserve(sectionRef.current, onInViewport);
  }, [sectionRef.current]);

  // Инициируем загрузку данных, когда компонент почти попал в зону видимости
  useEffect(() => {
    isFunction(addObserveWithMargin)
    && isFunction(onRequest)
    && emptyRef.current
    && addObserveWithMargin(emptyRef.current, onRequest);
  }, [emptyRef.current]);

  return isNonEmptyArray(items)
    ? (
      <section
        className={cx('wrapper')}
        ref={sectionRef}
      >
        <h3 className={cx('title')}>{title}</h3>
        <Carousel
          items={items}
          renderItem={(item, key) => (
            <RecommendedItem
              {...item}
              onItemClick={() => isFunction(onItemClick) && onItemClick(item)}
              key={key}
            />
          )}
          renderControl={CarouselButton}
        />
      </section>
    )
    : <div ref={emptyRef}></div>;
};

RecommendationCarousel.propTypes = {
  /**
   * Заголовок карусели.
   */
  title: Types.string,

  /**
   * Массив элементов карусели.
   */
  items: Types.arrayOf(Types.shape({
    name: Types.string,
    imageSrc: Types.string,
    imageAlt: Types.string,
    url: Types.string,
    price: Types.number,
    currencyGrapheme: Types.string,
    onQuickViewClick: Types.func,
    oldPrice: Types.number,
  })),

  /**
   * Обработчик клика на элементе карусели.
   */
  onItemClick: Types.func,

  /**
   * Функция, будет вызвана при попадании карусели во вьюпорт.
   */
  onInViewport: Types.func,

  /**
   * Функция для получения элементов карусели, будет вызвана при приближении карусели к вьюпорту.
   */
  onRequest: Types.func,

  /**
   * Функция подписки на Intersection Observer, предоставляется HOC'ом.
   */
  addObserve: Types.func,

  /**
   * Функция подписки на Intersection Observer, предоставляется HOC'ом.
   */
  addObserveWithMargin: Types.func,
};

export default withInViewportObserver(
  withInViewportObserver(RecommendationCarousel), {
    rootMargin: '200px 0px 200px 0px',
  }, 'addObserveWithMargin'
);
