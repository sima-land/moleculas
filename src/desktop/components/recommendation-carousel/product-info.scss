.row + .row {
  margin-top: 8px;
}

.image-row {
  position: relative;
  display: block;
  width: 100%;
  .image {
    width: 100%;
  }
  .quick-view-btn {
    position: absolute;
    right: 8px;
    top: 8px;
  }
}

.badges-row {
  display: flex;
  flex-wrap: wrap;
  max-height: 56px;
  overflow: hidden;
  margin-bottom: -8px;
  .badge {
    margin: 0 4px 8px 0;
  }
}

.prices-row {
  display: flex;
  align-items: center;
  height: 24px;
  .price {
    font-size: 16px;
    line-height: 24px;
    font-weight: 700;
  }
  .old-price {
    font-size: 12px;
    line-height: 16px;
    margin-left: 8px;
    font-weight: 400;
  }
}

.link {
  overflow: hidden;
  word-break: break-word;
  max-width: 100%;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
}
