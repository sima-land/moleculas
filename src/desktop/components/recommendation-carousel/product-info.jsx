import React from 'react';
import { Link } from '@dev-dep/ui-nucleons/link';
import { Price } from '@dev-dep/ui-nucleons/price';
import { Image } from '../../../common/components/image';
import { isNonEmptyArray } from '../../../common/helpers/is-non-empty';
import { Badge } from '../../../common/components/badge';
import classnames from 'classnames/bind';
import styles from './product-info.scss';
import { WishButton } from '../../../common/components/wish-button';
import { QuickViewButton } from '../../../common/components/quick-view-button';
import withPrevent from '@dev-dep/ui-nucleons/helpers/with-prevent';

const cx = classnames.bind(styles);

/**
 * Компонент базовой информации о товаре.
 * @param {Object} props Свойства компонента.
 * @return {ReactElement} Элемент.
 */
export const ProductInfo = ({
  badges,
  currencyGrapheme,
  favorite,
  imageSrc,
  name,
  oldPrice,
  onFavoriteClick,
  onQuickViewClick,
  price,
  url,
  withImageButtons,
}) => (
  <>
    <a className={cx('row', 'image-row')} href={url}>
      <Image
        className={cx('image')}
        alt={name}
        src={imageSrc}
      />
      {withImageButtons && (
        <>
          <WishButton
            checked={favorite}
            className={cx('on-image-btn', 'wish')}
            onClick={withPrevent(onFavoriteClick)}
          />
          <QuickViewButton
            className={cx('on-image-btn', 'quick-view')}
            onClick={withPrevent(onQuickViewClick)}
          />
        </>
      )}
    </a>

    {Boolean(price) && (
      <div className={cx('row', 'prices-row')}>
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price')}
        />
        {Boolean(oldPrice) && (
          <Price
            value={oldPrice}
            currencyGrapheme={currencyGrapheme}
            className={cx('old-price')}
            crossedOut
          />
        )}
      </div>
    )}

    {isNonEmptyArray(badges) && (
      <div className={cx('row', 'badges-row')}>
        {badges.map((badge, index) => (
          <Badge
            {...badge}
            key={index}
            className={cx('badge')}
          />
        ))}
      </div>
    )}

    <Link
      className={cx('row', 'link')}
      href={url}
      children={name}
      color='gray87'
    />
  </>
);
