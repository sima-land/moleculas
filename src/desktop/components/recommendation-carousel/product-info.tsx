import React from 'react';
import { Link } from '@dev-dep/ui-nucleons/link';
import { Price } from '@dev-dep/ui-nucleons/price';
import { Image } from '../../../common/components/image';
import { Badge } from '../../../common/components/badge';
import classnames from 'classnames/bind';
import styles from './product-info.scss';
import { QuickViewButton } from '../../../common/components/quick-view-button';
import withPrevent from '@dev-dep/ui-nucleons/helpers/with-prevent';
import { WithHint } from '@dev-dep/ui-nucleons/with-hint';
import { Product } from './types';

export interface ProductInfoProps extends Product {
  onQuickViewClick?: React.MouseEventHandler<SVGSVGElement>
  withQuickView?: boolean
}

const cx = classnames.bind(styles);

/**
 * Компонент базовой информации о товаре.
 * @param props Свойства компонента.
 * @return Элемент.
 */
export const ProductInfo = ({
  badges,
  currencyGrapheme,
  imageSrc,
  name,
  oldPrice,
  onQuickViewClick,
  price,
  url,
  withQuickView,
}: ProductInfoProps) => (
  <>
    <a className={cx('row', 'image-row')} href={url}>
      <Image
        className={cx('image')}
        alt={name}
        src={imageSrc}
      />
      {withQuickView && (
        <WithHint hint='Быстрый просмотр' direction='left'>
          {(ref, toggle) => (
            <QuickViewButton
              ref={ref as any}
              onMouseEnter={() => toggle(true)}
              onMouseLeave={() => toggle(false)}
              className={cx('quick-view-btn')}
              onClick={withPrevent(onQuickViewClick)}
            />
          )}
        </WithHint>
      )}
    </a>

    {Boolean(price) && (
      <div className={cx('row', 'prices-row')}>
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price')}
        />

        {oldPrice && (
          <Price
            value={oldPrice}
            currencyGrapheme={currencyGrapheme}
            className={cx('old-price')}
            crossedOut
          />
        )}
      </div>
    )}

    {Array.isArray(badges) && badges.length > 0 && (
      <div className={cx('row', 'badges-row')}>
        {badges.map((badge, index) => (
          <Badge
            {...badge}
            key={index}
            className={cx('badge')}
          />
        ))}
      </div>
    )}

    <Link
      className={cx('row', 'link')}
      href={url}
      children={name}
      color='gray87'
    />
  </>
);
