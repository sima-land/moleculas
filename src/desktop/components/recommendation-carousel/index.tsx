import React, { useRef, useState } from 'react';
import { Carousel } from '@sima-land/ui-nucleons/carousel';
import { ProductInfo } from './product-info';
import { HoverCard, HoverCardProps } from './hover-card';
import { useAllowFlag, useChildWidth, useViewport } from './utils';
import { useMedia } from '@sima-land/ui-nucleons/hooks/media';
import { CarouselItem } from './types';
import classnames from 'classnames/bind';
import styles from './recommendation-carousel.module.scss';

interface ItemSize {
  xs?: 2 | 3 | 4
  s?: 2 | 3 | 4
  m?: 2 | 3 | 4
  l?: 2 | 3 | 4
  xl?: 2 | 3 | 4
}

export interface RecommendationCarouselProps {
  className?: string
  items: CarouselItem[]
  itemSize?: ItemSize
  onInViewport?: () => void
  onItemAdd?: HoverCardProps['onAdd']
  onItemChange?: HoverCardProps['onChange']
  onItemQuickViewClick?: HoverCardProps['onQuickViewClick']
  onItemSubtract?: HoverCardProps['onSubtract']
  onNeedRequest?: () => void
  withHoverCard?: boolean
  title?: string
  titleTag?: 'div' | 'header' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
}

const cx = classnames.bind(styles);

// eslint-disable-next-line require-jsdoc
const getSizeClasses = (size = {}) => [
  `size-xs-${(size as any).xs || 4}`,
  `size-s-${(size as any).s || 3}`,
  `size-m-${(size as any).m || 2}`,
  `size-l-${(size as any).l || 2}`,
  `size-xl-${(size as any).xl || 2}`,
];

/**
 * Карусель рекомендованных товаров.
 * @param props Свойства.
 * @return Элемент.
 */
export const RecommendationCarousel = ({
  className,
  items,
  itemSize,
  onInViewport,
  onItemAdd,
  onItemChange,
  onItemQuickViewClick,
  onItemSubtract,
  onNeedRequest,
  withHoverCard,
  title,
  titleTag: Header = 'h2',
}: RecommendationCarouselProps) => {
  const [activeItem, setActiveItem] = useState<CarouselItem | null>(null);
  const cardShow = useAllowFlag();
  const needBigArrows = useMedia('(min-width: 1440px)');
  const sectionRef = useRef<HTMLElement | null>(null);
  const stubRef = useRef<HTMLDivElement | null>(null);
  const targetItemRef = useRef<HTMLDivElement | null>(null);

  // вычисляем ширину элемента карусели для позиционирования стрелок
  const itemWidth = useChildWidth(sectionRef, `.${cx('item')}`, [items]);

  // инициируем загрузку данных, когда компонент почти попал в зону видимости
  useViewport(stubRef, onNeedRequest);

  // отправляем статистку в ОКО, когда компонент попадает в зону видимости
  useViewport(sectionRef, onInViewport, {
    rootMargin: '200px 0px 200px 0px',
  });

  return Array.isArray(items) && items.length > 0
    ? (
      <section ref={sectionRef} className={cx('root', className)}>
        <Header className={cx('title')}>{title}</Header>

        <Carousel step={3}
          draggable={false}
          items={items}
          withControls={itemWidth !== null}
          controlProps={{
            size: needBigArrows ? 'l' : 's',
            style: {
              zIndex: 1,

              // @ts-expect-error ts-migrate(2531) FIXME: Object is possibly 'null'.
              top: `${itemWidth / 2}px`,
            },
          }}
          renderItem={(item: CarouselItem) => (
            <div
              data-testid='reco-item'
              className={cx('item', getSizeClasses(itemSize))}
              onMouseEnter={e => {
                if (cardShow.allowed()) {
                  targetItemRef.current = e.currentTarget;
                  setActiveItem(item);
                }
              }}
            >
              <ProductInfo {...item.product} withQuickView={false} />
            </div>
          )}

          // длительность прокрутки в Carousel - 320, делаем слегка с запасом
          // @todo после восстановления проверить позицию курсора чтобы показать карточку (если будет критично)
          onChangeTargetIndex={() => {
            cardShow.disallowFor(360);
          }}
        />

        {withHoverCard && activeItem && (
          <HoverCard
            info={activeItem}
            targetRef={targetItemRef}
            onQuickViewClick={onItemQuickViewClick}
            onAdd={onItemAdd}
            onSubtract={onItemSubtract}
            onChange={onItemChange}
            onMouseLeave={() => setActiveItem(null)}
          />
        )}
      </section>
    )
    : (
      <div ref={stubRef} />
    );
};
