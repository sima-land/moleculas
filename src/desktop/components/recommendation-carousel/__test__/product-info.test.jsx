import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { ProductInfo } from '../product-info';
import { QuickViewButton } from '../../../../common/components/quick-view-button';

describe('<ProductInfo />', () => {
  it('should renders correctly', () => {
    const wrapper = mount(
      <ProductInfo
        url='https://www.sima-land.ru/123456'
        name='Ножницы портновские, 10, 26 см, цвет чёрный'
        imageSrc='https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476'
        price={225}
        oldPrice={250}
        favorite
        currencyGrapheme='₽'
        withQuickView
        onQuickViewClick={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "onLinkClick" prop', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <ProductInfo
        url='https://www.sima-land.ru/123456'
        name='Ножницы портновские, 10, 26 см, цвет чёрный'
        imageSrc='https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476'
        price={225}
        currencyGrapheme='₽'
        onLinkClick={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find('a[data-testid="product-info:image-link"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(1);

    act(() => {
      wrapper.find('a[data-testid="product-info:name-link"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(2);
  });

  it('should renders hints properly', () => {
    const wrapper = mount(
      <ProductInfo
        url='https://www.sima-land.ru/123456'
        name='Ножницы портновские, 10, 26 см, цвет чёрный'
        imageSrc='https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476'
        price={225}
        oldPrice={250}
        favorite
        currencyGrapheme='₽'
        withQuickView
        onQuickViewClick={jest.fn()}
      />
    );

    act(() => {
      wrapper.find(QuickViewButton).simulate('mouseenter');
    });
    wrapper.update();

    expect(wrapper.find('[data-testid="hint"]')).toHaveLength(1);

    act(() => {
      wrapper.find(QuickViewButton).simulate('mouseleave');
    });
    wrapper.update();

    expect(wrapper.find('[data-testid="hint"]')).toHaveLength(0);
  });
});
