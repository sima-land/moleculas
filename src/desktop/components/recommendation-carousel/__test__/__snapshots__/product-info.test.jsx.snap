// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductInfo /> should renders correctly 1`] = `
<ProductInfo
  currencyGrapheme="₽"
  favorite={true}
  imageSrc="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476"
  name="Ножницы портновские, 10, 26 см, цвет чёрный"
  oldPrice={250}
  onFavoriteClick={[MockFunction]}
  onQuickViewClick={[MockFunction]}
  price={225}
  url="https://www.sima-land.ru/123456"
  withImageButtons={true}
>
  <a
    className="row image-row"
    href="https://www.sima-land.ru/123456"
  >
    <Image
      alt="Ножницы портновские, 10, 26 см, цвет чёрный"
      className="image"
      src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476"
    >
      <div
        className="root image"
      >
        <img
          alt="Ножницы портновские, 10, 26 см, цвет чёрный"
          src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v=1617955476"
        />
      </div>
    </Image>
    <WishButton
      className="on-image-btn wish"
      isWished={true}
      onClick={[Function]}
    >
      <div
        className="wish-button on-image-btn wish is-wished"
        data-testid="favorite-button"
        onClick={[Function]}
        title="Убрать из избранного"
      >
        <HeartFilledSVG
          className="svg stroke"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <svg
            className="svg stroke"
            height={24}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.657 4.01C17.018 2.174 13.76 3.031 12 5.132c-1.759-2.1-5.018-2.968-7.657-1.122-1.4.98-2.279 2.632-2.339 4.376-.14 3.958 3.299 7.13 8.547 11.996l.1.091c.76.704 1.93.704 2.689-.01l.11-.102c5.248-4.855 8.676-8.027 8.546-11.985-.06-1.734-.94-3.386-2.339-4.366z"
            />
          </svg>
        </HeartFilledSVG>
      </div>
    </WishButton>
    <QuickViewButton
      className="on-image-btn quick-view"
      onClick={[Function]}
    >
      <div
        className="button on-image-btn quick-view"
        data-testid="quick-view-button"
        onClick={[Function]}
        title="Быстрый просмотр"
      >
        <QuickViewSVG
          className="stroke"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <svg
            className="stroke"
            height={24}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.94 22.06l-3.219-3.217-.339.22A8.956 8.956 0 0 1 11.5 20.5a9 9 0 1 1 9-9c0 1.8-.528 3.476-1.438 4.882l-.22.34 3.219 3.217a1.5 1.5 0 0 1-2.122 2.122zM9 10h1V9a1.5 1.5 0 1 1 3 0v1h1a1.5 1.5 0 1 1 0 3h-1v1a1.5 1.5 0 1 1-3 0v-1H9a1.5 1.5 0 1 1 0-3zm2.5-4.5a6 6 0 1 0 0 12 6 6 0 0 0 0-12z"
            />
          </svg>
        </QuickViewSVG>
      </div>
    </QuickViewButton>
  </a>
  <div
    className="row prices-row"
  >
    <Price
      className="price"
      currencyGrapheme="₽"
      value={225}
    >
      <span
        className="root price"
      >
        225 ₽
      </span>
    </Price>
    <Price
      className="old-price"
      crossedOut={true}
      currencyGrapheme="₽"
      value={250}
    >
      <span
        className="root old-price crossed-out"
      >
        250 ₽
      </span>
    </Price>
  </div>
  <ForwardRef(Link)
    className="row link"
    color="gray87"
    href="https://www.sima-land.ru/123456"
  >
    <a
      className="link row link color__gray87 hover-color__gray54"
      href="https://www.sima-land.ru/123456"
    >
      Ножницы портновские, 10, 26 см, цвет чёрный
    </a>
  </ForwardRef(Link)>
</ProductInfo>
`;
